Re-sync hwb() implementation with the latest spec

weinig opened the pull request:
#### 13c529c28ace81577cfeec0ce964d4d6f31522ad
<pre>
Re-sync hwb() implementation with the latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=272984">https://bugs.webkit.org/show_bug.cgi?id=272984</a>

Reviewed by Darin Adler.

The CSS Color specification has changed since hwb() was last
worked on. This updates things to support:

- Modified parsing rules to normalize hue and allow any value for whiteness/blackness,
  removing all eager whiteness/blackness normalization.
- Changes to the types used by the h/w/b/alpha symbols used in the relative
  color form to always be CSS_NUMBER.
- Requirement that the relative form serialize as `color(srgb )`.
- Updated conversion rules to account for unbounded range and related changes to HSL
  conversion.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-hwb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    Update another good chunk of tests to now pass.

* Source/WebCore/platform/graphics/ColorNormalization.h:
(WebCore::makeCanonicalColor): Deleted.
(WebCore::normalizeClampedWhitenessBlacknessDisallowingNone): Deleted.
(WebCore::normalizeClampedWhitenessBlacknessAllowingNone): Deleted.
(WebCore::normalizeWhitenessBlackness): Deleted.
(WebCore::makeColorTypeByNormalizingComponents&lt;HWBA&lt;float&gt;&gt;): Deleted.
(WebCore::makeCanonicalColor&lt;HWBA&lt;float&gt;&gt;): Deleted.
* Source/WebCore/platform/graphics/ColorInterpolation.cpp:
(WebCore::interpolateColors):
* Source/WebCore/css/color/CSSResolvedColorMix.cpp:
(WebCore::mixColorComponentsUsingColorInterpolationMethod):
    Removed last specialization of makeCanonicalColor() and then remove all
    callers since it now is just a straight creation of Color.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::normalizeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw):
    Work to separate parsing and normalization, allowing much of normalization
    to be reused for all the parsing paths.

* Source/WebCore/platform/graphics/ColorTypes.h:
    Update HWBA to use ExtendedSRGBA as its reference (like HSLA).

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::ExtendedSRGBA&lt;float&gt;&gt;::convert):
(WebCore::HWBA&lt;float&gt;&gt;::convert):
(WebCore::calculateHSLHue): Deleted.
(WebCore::SRGBA&lt;float&gt;&gt;::convert): Deleted.
* Source/WebCore/platform/graphics/ColorConversion.h:
    Update conversion code to use new reference type and updated
    algorithm.

* Source/WebCore/platform/graphics/ColorModels.h:
    Update component information to new constraints.

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::serializationForCSS):
    Ensure we use all the precision available when serializing
    as color(srgb) for HWBA by converting to ExtendedSRGBA instead
    of the bounded SRGBA. Update HSLA as well, since that has the
    same problem.

Canonical link: <a href="https://commits.webkit.org/277999@main">https://commits.webkit.org/277999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6f2e8d48e5b48dfa01f37428e3c72b905b328ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49143 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45239 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51447 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25998 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40170 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50821 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42362 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23439 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43536 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7467 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45367 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44043 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53875 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24221 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20426 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47506 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42567 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10818 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26316 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25218 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -820,96 +820,134 @@ static Color parseHSLParametersRaw(CSSParserTokenRange& range, ColorParserState&
+    return HWBA<float> {
+        static_cast<float>(normalizedHue),
+        static_cast<float>(normalizedWhiteness),
+        static_cast<float>(normalizedBlackness),
```
We could just use `float` rather than `auto` above for these local variables, and avoid the need for these 3 casts.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -820,96 +820,134 @@ static Color parseHSLParametersRaw(CSSParserTokenRange& range, ColorParserState&
+        [] (NoneRaw) { return std::numeric_limits<double>::quiet_NaN(); }
+    );
+
+    return HWBA<float> {
```
Can we omit the template class name here and just use braces, since the return type is `HWBA<float>`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -820,96 +820,134 @@ static Color parseHSLParametersRaw(CSSParserTokenRange& range, ColorParserState&
-    // that HWBA colors serialize using the legacy color syntax (rgb()/rgba()) and historically have used the 8-bit rgba
-    // internal representation in engines.
-    return convertColor<SRGBA<uint8_t>>(HWBA<float> { static_cast<float>(normalizedHue), static_cast<float>(normalizedWhitness), static_cast<float>(normalizedBlackness), static_cast<float>(*alpha) });
+    return ParsedHWBA { *hue, *whiteness, *blackness, *alpha };
```
Is there an alternate here where we just use two sets of braces instead of naming the type?

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -154,63 +126,75 @@ ExtendedSRGBA<float> ColorConversion<ExtendedSRGBA<float>, HSLA<float>>::convert
+        return (component * whitenessBlacknessFactor) + scaledWhiteness;
     };
+
+    return { hueToRGB(0.0f), hueToRGB(8.0f), hueToRGB(4.0f), alpha };
```
I’d be so tempted to just write `0`, `8`, and `4` here.

darinadler commented on Source/WebCore/platform/graphics/ColorSerialization.cpp:
```diff
@@ -378,6 +378,9 @@ String serializationForRenderTreeAsText(const ExtendedSRGBA<float>& color, bool)
+    if (useColorFunctionSerialization)
+        return serializationForCSS(convertColor<ExtendedSRGBA<float>>(color), useColorFunctionSerialization);
+
     return serializationForCSS(convertColor<SRGBA<uint8_t>>(color), useColorFunctionSerialization);
```
A little strange to be passing the boolean in for functions on both sides of the `if` statement. I’d be tempted to write `true` and `false`.

weinig replied:
Yeah, that is dumb. Will fix there and in a few other places.

darinadler commented on Source/WebCore/platform/graphics/ColorSerialization.cpp:
```diff
@@ -395,6 +398,9 @@ String serializationForRenderTreeAsText(const HSLA<float>& color, bool useColorF
+    if (useColorFunctionSerialization)
+        return serializationForCSS(convertColor<ExtendedSRGBA<float>>(color), useColorFunctionSerialization);
+
     return serializationForCSS(convertColor<SRGBA<uint8_t>>(color), useColorFunctionSerialization);
```
A little strange to be passing the boolean in for functions on both sides of the `if` statement. I’d be tempted to write `true` and `false`.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
```diff
@@ -99,8 +99,14 @@ PASS e.style['color'] = "color-mix(in hwb, hwb(120deg 10% 20% / .4) 25%, hwb(30d
 PASS e.style['color'] = "color-mix(in hwb, hwb(120deg 10% 20% / .4) 0%, hwb(30deg 30% 40% / .8))" should set the property value
-PASS e.style['color'] = "color-mix(in hwb, hwb(40deg 30% 40%), hwb(60deg 30% 40%))" should set the property value
-PASS e.style['color'] = "color-mix(in hwb, hwb(60deg 30% 40%), hwb(40deg 30% 40%))" should set the property value
+FAIL e.style['color'] = "color-mix(in hwb, hwb(40deg 30% 40%), hwb(60deg 30% 40%))" should set the property value Colors do not match.
+Actual:   color-mix(in hwb, rgb(153, 127, 77), rgb(153, 153, 77))
+Expected: color-mix(in hwb, rgb(153, 128, 77), rgb(153, 153, 77)).
+Error: assert_array_approx_equals: Numeric parameters are approximately equal. property 1, expected 128 +/- 0.01, expected 128 but got 127
+FAIL e.style['color'] = "color-mix(in hwb, hwb(60deg 30% 40%), hwb(40deg 30% 40%))" should set the property value Colors do not match.
+Actual:   color-mix(in hwb, rgb(153, 153, 77), rgb(153, 127, 77))
+Expected: color-mix(in hwb, rgb(153, 153, 77), rgb(153, 128, 77)).
+Error: assert_array_approx_equals: Numeric parameters are approximately equal. property 4, expected 128 +/- 0.01, expected 128 but got 127
```
What should I think about all the new 127 vs. 128 failures in this file?

mdubet replied:
We don't round 127.5 correctly https://bugs.webkit.org/show_bug.cgi?id=256888

weinig replied:
This comes down to a small shift when doing the math where we get 0.500000 as the green value that when scaled to be out of 255 becomes 127.499985 and gets rounded down.

We currently use use `round`, but I think the CSS spec wants us to `ceil`:

> The percentage reference range of the color channels comes from the historical fact that many graphics engines stored the color channels internally as a single byte, which can hold integers between 0 and 255. Implementations should honor the precision of the channel as authored or calculated wherever possible. If this is not possible, the channel should be [rounded towards +∞](https://drafts.csswg.org/css-values-4/#combine-integers).

The linked text at the end unfortunately kind of contradicts that, saying only rounding towards +∞ when exactly 0.5.

Will be following up on this, but probably the test should be updated to allow +/- 1, not +/- 0.01 since these are integers.

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -824,96 +824,134 @@ static Color parseHSLParametersRaw(CSSParserTokenRange& range, ColorParserState&
     auto hue = hueConsumer(args);
     if (!hue)
-        return { };
+        return std::nullopt;
```
We should keep the more generic uniform initialization `{ }`

Commit:
Re-sync hwb() implementation with the latest spec
https://bugs.webkit.org/show_bug.cgi?id=272984

Reviewed by Darin Adler.

The CSS Color specification has changed since hwb() was last
worked on. This updates things to support:

- Modified parsing rules to normalize hue and allow any value for whiteness/blackness,
  removing all eager whiteness/blackness normalization.
- Changes to the types used by the h/w/b/alpha symbols used in the relative
  color form to always be CSS_NUMBER.
- Requirement that the relative form serialize as `color(srgb )`.
- Updated conversion rules to account for unbounded range and related changes to HSL
  conversion.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-hwb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    Update another good chunk of tests to now pass.

* Source/WebCore/platform/graphics/ColorNormalization.h:
(WebCore::makeCanonicalColor): Deleted.
(WebCore::normalizeClampedWhitenessBlacknessDisallowingNone): Deleted.
(WebCore::normalizeClampedWhitenessBlacknessAllowingNone): Deleted.
(WebCore::normalizeWhitenessBlackness): Deleted.
(WebCore::makeColorTypeByNormalizingComponents<HWBA<float>>): Deleted.
(WebCore::makeCanonicalColor<HWBA<float>>): Deleted.
* Source/WebCore/platform/graphics/ColorInterpolation.cpp:
(WebCore::interpolateColors):
* Source/WebCore/css/color/CSSResolvedColorMix.cpp:
(WebCore::mixColorComponentsUsingColorInterpolationMethod):
    Removed last specialization of makeCanonicalColor() and then remove all
    callers since it now is just a straight creation of Color.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::normalizeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw):
    Work to separate parsing and normalization, allowing much of normalization
    to be reused for all the parsing paths.

* Source/WebCore/platform/graphics/ColorTypes.h:
    Update HWBA to use ExtendedSRGBA as its reference (like HSLA).

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::ExtendedSRGBA<float>>::convert):
(WebCore::HWBA<float>>::convert):
(WebCore::calculateHSLHue): Deleted.
(WebCore::SRGBA<float>>::convert): Deleted.
* Source/WebCore/platform/graphics/ColorConversion.h:
    Update conversion code to use new reference type and updated
    algorithm.

* Source/WebCore/platform/graphics/ColorModels.h:
    Update component information to new constraints.

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::serializationForCSS):
    Ensure we use all the precision available when serializing
    as color(srgb) for HWBA by converting to ExtendedSRGBA instead
    of the bounded SRGBA. Update HSLA as well, since that has the
    same problem.

Canonical link: https://commits.webkit.org/277999@main
